apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo/network/ws/ApolloWebSocketNetworkTransport.kt

```diff
@@ -205,22 +214,34 @@ class ApolloWebSocketNetworkTransport(
 
     private suspend fun onUnsubscribed() {
       mutex.withLock {
-        if (--activeSubscriptionCount == 0) scheduleAutoClose()
+        if (--activeSubscriptionCount == 0 && idleTimeoutMs > 0) {
+          idleTimeoutJob?.cancel()
+          connectionKeepAliveTimeoutJob?.cancel()
+
+          idleTimeoutJob = coroutineScope.launch {
+            delay(idleTimeoutMs)
+            close()
+          }
+        }
       }
     }
 
-    private fun scheduleAutoClose() {
-      idleTimeoutJob?.cancel()
-      if (idleTimeoutMs > 0) {
-        idleTimeoutJob = coroutineScope.launch {
-          delay(idleTimeoutMs)
-          close()
+    private suspend fun onConnectionKeepAlive() {
+      mutex.withLock {
+        if (activeSubscriptionCount > 0 && connectionKeepAliveTimeoutMs > 0) {
+          connectionKeepAliveTimeoutJob?.cancel()
+
+          connectionKeepAliveTimeoutJob = coroutineScope.launch {
+            delay(idleTimeoutMs)
+            close()
+          }
         }
       }
     }
 
     private suspend fun close() {
       mutex.withLock {
+        yield()
```

Comment:
Why is this `yield` required?

Reply:
After second thought seems there is no need for it here.

My initial thinking was that the coroutine might have been canceled before we acquire mutex, so to handle this case we call `yield ` that will fail if coroutine is canceled. But `mutex.withLock ` is suspended function so it will be handled anyway, if coroutine is canceled then `mutex.withLock` will be canceled.